netlogo-core/src/main/api/FileIO.scala

```diff
@@ -124,6 +124,33 @@ object FileIO {
     finally file.close(false)
   }
 
+  def perUserFile(file: String, createNecessaryDirs: Boolean = true): String = {
+    val res = perUserPath(file)
+    if (createNecessaryDirs)
+      Files.createDirectories(Paths.get(res).getParent)
+    res
+  }
+
+  def perUserDir(dir: String, create: Boolean = true): String = {
+    val res = perUserPath(dir)
+    if (create)
+      Files.createDirectories(Paths.get(res))
+    res
+  }
+
+  private def perUserPath(path: String): String = {
+    val os = System.getProperty("os.name").toUpperCase
+    val appData =
+      if (os.contains("WIN"))
+        System.getenv("APPDATA") + "\\NetLogo"
+      else if (os.contains("MAC"))
+        System.getProperty("user.home") + "/Library/Application Support/NetLogo"
+      else
+        System.getProperty("user.home") + "/.netlogo"
+
+    appData + File.separator + path
+  }
+
```

Comment:
Why is all of this being exposed in the NetLogo API?  That doesn't seem right to me.

Reply:
I agree, but this package already contain a lot of classes (or traits) that aren't really part of the NetLogo API, but are instead internal to NetLogo, and I couldn't think of a better place.

Reply:
Well, if there are a bunch of things in `api` that aren't really a part of the API, that seems like a negative thing to me, and I feel that we shouldn't actively make the problem worse.  However, I _can_ be convinced that this is our best reasonable move.

Can you explain why it's better for this to be in `api` than in `util.Utils`, where you had it before?

Reply:
Per the Package Guide in the wiki:
>  nothing in here is supposed to be NetLogo-specific

Looking at the other functions there, this rule is being adhered to.

Reply:
Okay.  I don't really like it, but the argument against putting it in the `util` package is clear, and `RemoteFile` and `StringUtils` _do_ set a precedent for dumping this sort of stuff in `api` (much as I might disagree).